### PR TITLE
Refactor/ simplify long polling

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

My first try to improve our current situation and to make it a bit more understandable, of course, this is discussable whether it helps. Happy to hear your feedback (since I'm now biased as I know now what the code does).

PR includes 

- Several methods have been renamed to better reflect what they do
- Some methods have been inlined to reduce complexity and better visibility
- Some method calls have been decoupled to reduce loops and reduce cognitive load


Be aware that I haven't changed any logic yet, it is a simple refactoring. My plan is to clean up also the multiple subsequent actor calls which I see as unnecessary and making it harder to understand (furthermore it might be delaying things as well).


Furthermore, I added a comment to explain the interactions and "state machine"

```
                                                                                                                                                                                                                                                           
                                                +--------------------------------------------------------------------------------+                                                                                                                         
                                                |                                   BROKER(s)                                    |                                                                                                                         
                                                +--------|-------------------------------^---------------------------------------+                                                                                                                         
                                                         |                               |                                                                                                                                                                 
+-----------+      +-------------------------------------|-------------------------------|--------------------------------------------------------------------------------+                                                                                
|           |      |                                     |          GATEWAY              | Activate commands/responses                                                    |                                                                                
|           |      |                                     |                               |                                                                                |                                                                                
|           |      |                                     |                               |                                                                                |                                                                                
|           |      |                                     |                               |                                                                                |                                                                                
|           |      |   +-----------------------+         |          +--------------------v-------------------+ No Jobs   +-------------------------------+                |                                                                                
|         --|------|--->   Incoming Activate   ----------|---------->   Try to Activate on all partitions    ------------|   Handle No Jobs Activated    |                |                                                                                
|           |      |   +-----------------------+         |          +---------|----------^-------------------+           +-------/---------------|-------+                |                                                                                
|           |      |                        +------------------------+        |          |                                      |                |                        |                                                                                
|           |      |               +---------  Jobs available        |        |          |                                      /                |                        |                                                                                
|           |      |               |        |  Notification          |        |          |                    +----------------v-----+           |                        |                                                                                
|           |      |               |        +------------------------+        |          |               +---->Complete Or Resubmit  |           |                        |                                                                                
|           |      |               |                                          |          |               |    +-----|----------|-----+           |                        |                                                                                
| Client  <-|------|--------+--------------------------+----------------------+          |               |          |          |                 |                        |                                                                                
|           |      |        |                          |        Received Jobs            |               |          |          |                 |                        |                                                                                
|           |      |        | Handle Pending Requests  |                                  \              |          | +--------v-------+         |   Any request          |                                                                                
|           |      |        |                          |                                  |     Reached  |          | |  Mark Pending  |         |   contained            |                                                                                
|           |      |        +--------------------------+                                  |     failure  |          | +----------------+         |   resource exhausted   |                                                                                
|           |      |                      |                                               |     threshold|          |                            |                        |                                                                                
|           |      |                      |                                               |              |          |                            |                        |                                                                                
|           |      |                      |                                               |              |          |Retrieved                   |                        |                                                                                
|           |      |                      |                                               |              |          |notify                      |                        |                                                                                
|           |      |                      |  For all pending requests                     |              |          |in between                  |                        |                                                                                
|           |      |                      |                                       +-------|--------------|-------+  |                            |                        |                                                                                
|           |      |                      +---------------------------------------> Internal Activate Jobs Retry <--+                            |                        |                                                                                
|           |      |                                                              +------------------------------+                               |                        |                                                                                
|           |      |                                                                                                                             |                        |                                                                                
|           |------|-----------                                                                                                                  |                        |                                                                                
+-----------+      |           \-------------------------------------                                                         +------------------v-----+                  |                                                                                
                   |                                         Error   \-------------------------------------                   |                        |                  |                                                                                
                   |                                                                                       \-------------------   Resource Exhausted   |                  |                                                                                
                   |                                                                                                          |                        |                  |                                                                                
                   |                                                                                                          +------------------------+                  |                                                                                
                   |                                                                                                                                                      |                                                                                
                   |                                                                                                                                                      |                                                                                
                   |                                                                                                                                                      |                                                                                
                   +------------------------------------------------------------------------------------------------------------------------------------------------------+                                                                                
                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                 
```

https://textik.com/#a2725e317ed87a9d
<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

Related to a recent investigation for a support case, to make it clearer for next time.

\cc @npepinpe maybe you are interested as well.

### CODE REVIEW

**Comment by Zelldon** [comment]
I would like to backport these changes, so we don't have to much issues with backporting other changes

**Review by lenaschoenburg — COMMENTED** [review]
I read through and I think I like the changes. The diagram is also super useful :+1: 

I haven't reviewed it critically yet and since long-polling is complicated I'd like to do another review tomorrow, unless @npepinpe also reviews until then.

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java [hunk 1]
```diff
@@ -74,21 +72,140 @@ public void accept(final ActorControl actor) {
...
   public void activateJobs(
       final ActivateJobsRequest request,
       final ServerStreamObserver<ActivateJobsResponse> responseObserver) {
-    final var longPollingRequest = toInflightActivateJobsRequest(request, responseObserver);
-    activateJobs(longPollingRequest);
+    actor.run(
+        () -> {
+          final var longPollingRequest =
+              new InflightActivateJobsRequest(
+                  ACTIVATE_JOBS_REQUEST_ID_GENERATOR.getAndIncrement(), request, responseObserver);
```
Comment (npepinpe) [inline gateway/src/main/java/io/camunda/zeebe/gateway/impl/job/LongPollingActivateJobsHandler.java:135]: :wrench: If `ACTIVATE_JOBS_REQUEST_ID_GENERATOR` is only accessed through the actor, we don't need to make it an atomic.

**Review by npepinpe — APPROVED** [review]
Looks good to me generally, I think we could go even further into refactoring it. Now that we have a good model of what we want, we can shape the code to fit it (instead of using the code to discover the model, if that makes sense :smile:)

There's still even too many lambdas with 5-6 lines for my taste, but readability at that point is subjective, so it's fine. Maybe I'm just lambda averse :smile:

**Comment by Zelldon** [comment]
Yes, it makes total sense. As I wrote, it was for me the first increment where I was able to stop and get some feedback. To have quick feedback on whether im on the right track.

I also mentioned that i see no point in having all the actor calls, which would remove a lot of these lambdas as well. But I wanted to do this separately.

**Comment by Zelldon** [comment]
I updated the doc, to add the job notification.

**Pull request merged**